File: repository/shapes/cpp/Weirich/shapes.cc
/* Object Orientation in C++ */

#include <iostream.h>

/* abstract interface declaration */

class Shape
{
public:
    virtual void Draw () = 0;
    virtual void MoveTo (int newx, int newy) = 0;
    virtual void RMoveTo (int dx, int dy) = 0;
};
			 
/* Class Rectangle */

class Rectangle : public Shape
{
public:
    Rectangle (int x, int y, int w, int h);
    virtual void Draw ();
    virtual void MoveTo (int newx, int newy);
    virtual void RMoveTo (int dx, int dy);
    virtual void SetWidth (int newWidth);
    virtual void SetHeight (int newHeight);

private:
    int x, y;
    int width;
    int height;
};
    
void Rectangle::Draw ()
{
    cout << "Drawing a Rectangle at (" << x << "," << y
	 << "), width " << width << ", height " << height << "\n";
};

void Rectangle::MoveTo (int newx, int newy)
{
    x = newx;
    y = newy;
}

void Rectangle::RMoveTo (int dx, int dy)
{
    x += dx;
    y += dy;
}

void Rectangle::SetWidth (int newWidth)
{
    width = newWidth;
}

void Rectangle::SetHeight (int newHeight)
{
    height = newHeight;
}

Rectangle::Rectangle (int initx, int inity, int initw, int inith)
{
    x = initx;
    y = inity;
    width = initw;
    height = inith;
}


/* Class Circle */

class Circle : public Shape
{
public:
    Circle (int initx, int inity, int initr);
    virtual void Draw ();
    virtual void MoveTo (int newx, int newy);
    virtual void RMoveTo (int dx, int dy);
    virtual void SetRadius (int newRadius);

private:
    int x, y;
    int radius;
};

void Circle::Draw ()
{
    cout << "Drawing a Circle at (" << x << "," << y
	 << "), radius " << radius << "\n";
}

void Circle::MoveTo (int newx, int newy)
{
    x = newx;
    y = newy;
}

void Circle::RMoveTo (int dx, int dy)
{
    x += dx;
    y += dy;
}

void Circle::SetRadius (int newRadius)
{
    radius = newRadius;
}

Circle::Circle (int initx, int inity, int initr)
{
    x = initx;
    y = inity;
    radius = initr;
}


/* ===================================================================
 * DoSomethingWithShape is a fuction that takes a polymorphic shape
 * and manipulates it according to its interface.
 */

void DoSomethingWithShape (Shape * s)
{
    s->Draw ();
    s->RMoveTo (100, 100);
    s->Draw ();
}


/* =================================================================== 
 * Main Program
 */

int main ()
{
    
    /* using shapes polymorphically */

    Shape * shapes[2];
    shapes[0] = new Rectangle (10, 20, 5, 6);
    shapes[1] = new Circle (15, 25, 8);

    for (int i=0; i<2; ++i) {
	DoSomethingWithShape (shapes[i]);
    }

    /* access a rectangle specific function */

    Rectangle * rect = new Rectangle (0, 0, 15, 15);
    rect->SetWidth (30);
    rect->Draw ();

    return 0;
}

File: samples/Shapes.hs
{-# LANGUAGE EmptyDataDecls, DeriveDataTypeable, TemplateHaskell #-}
{-# OPTIONS_GHC -fcontext-stack=100 #-}

{-

OOHaskell (C) 2004--2007, Oleg Kiselyov, Ralf Laemmel

-- The classic Shapes example
-- http://www.angelfire.com/tx4/cus/shapes/
-- We shall try to emulate the `classical' C++ implementation
-- http://www.angelfire.com/tx4/cus/shapes/cpp.html

-}


module Shapes where

import OOHaskell

-- The class Shape

-- First, declare the labels.

$(label "getX")
$(label "getY")
$(label "setX")
$(label "setY")
$(label "moveTo")
$(label "moveBy")
$(label "draw")


-- Note that unlike C++ version, our class is polymorphic in x and y.
-- Those fields can be any Num.
-- The fields are private, just as they are in C++ code

shape x y self
  = do
       xRef <- newIORef x
       yRef <- newIORef y
       return $
            getX     .=. readIORef xRef
        .*. getY     .=. readIORef yRef
        .*. setX     .=. writeIORef xRef
        .*. setY     .=. writeIORef yRef
        .*. moveTo   .=. (\x y -> do (self # setX) x; (self # setY) y)
        .*. moveBy   .=. (\dx dy ->
              do
                 x  <- self # getX
                 y  <- self # getY
                 (self # moveTo) (x + dx) (y + dy))
        .*. emptyRecord


-- More labels

$(label "getWidth")
$(label "getHeight")
$(label "setWidth")
$(label "setHeight")


-- Rectange: inherits from Shape

rectangle x y width height self
  = do
       super     <- shape x y self
       widthRef  <- newIORef width
       heightRef <- newIORef height
       return $
            getWidth  .=. readIORef  widthRef
        .*. getHeight .=. readIORef  heightRef
        .*. setWidth  .=. writeIORef widthRef
        .*. setHeight .=. writeIORef heightRef
        .*. draw      .=. printLn ("Drawing a Rectangle at:("
                               << self # getX << "," << self # getY
                               << "), width " << self # getWidth
                               << ", height " << self # getHeight)
        .*. super


-- More labels

$(label "getRadius")
$(label "setRadius")


-- Circle: inherits from Shape

circle x y radius self
  = do
       super     <- shape x y self
       radiusRef <- newIORef radius
       return $
            getRadius  .=. readIORef radiusRef
        .*. setRadius  .=. writeIORef radiusRef
        .*. draw       .=. printLn ("Drawing a Circle at:("
                                << self # getX << "," << self # getY
                                << "), radius " << self # getRadius)
        .*. super


-- Test case for heterogeneous collections

main = do
          -- Construct a list of shapes
          s1 <- mfix $ rectangle 10 20 5 6
          s2 <- mfix $ circle 15 25 8
          let scribble = cons s1 (cons s2 nil)

          -- Handle the shapes in the list polymorphically       
          mapM_ (\s -> do
                          s # draw
                          (s # moveBy) 100 100
                          s # draw)
                scribble

          -- call a rectangle specific function
          arec <- mfix (rectangle (0::Int) (0::Int) 15 15)
          arec # setWidth $ 30
--          arec # setRadius $ 40
          arec # draw
